File: 13-RomanToInteger/Main.java
import java.util.Map;
import java.util.HashMap;

class RomanToInteger {
  public int romanToInteger(String s) {
    Map<Character, Integer> roman = new HashMap<>();
    roman.put('I', 1);
    roman.put('V', 5);
    roman.put('X', 10);
    roman.put('L', 50);
    roman.put('C', 100);
    roman.put('D', 500);
    roman.put('M', 1000);

    int total = 0;
    int length = s.length();

    for (int i = 0; i < length; i++) {
      int val = roman.get(s.charAt(i));
      if (i + 1 < length && val < roman.get(s.charAt(i + 1))) {
        total -= val;
      } else {
        total += val;
      }
    }

    return total;
  }

  public static void main(String[] args) {
    RomanToInteger sol = new RomanToInteger();
    System.out.println(sol.romanToInteger("III"));
    System.out.println(sol.romanToInteger("LVIII"));
    System.out.println(sol.romanToInteger("MCMXCIV"));
  }
}

File: 13-RomanToInteger/main.ts
const romanToInt = (s: string): number => {
  const roman: Map<string, number> = new Map([
    ["I", 1],
    ["V", 5],
    ["X", 10],
    ["L", 50],
    ["C", 100],
    ["D", 500],
    ["M", 1000],
  ]);

  let total: number = 0;
  const n: number = s.length;

  for (let i: number = 0; i < n; i++) {
    const val: number = roman.get(s[i])!;
    if (i + 1 < n && val < roman.get(s[i + 1])!) {
      total -= val;
    } else {
      total += val;
    }
  }

  return total;
};

console.log(romanToInt("III"));
console.log(romanToInt("LVIII"));
console.log(romanToInt("MCMXCIV"));
